corehq/motech/fhir/tasks.py

```diff
@@ -394,5 +398,252 @@ def create_parent_indices(
+    }
+    data = {
+        "grant_type": "client_credentials",
+        "client_assertion_type": "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
+        "client_assertion": generate_epic_jwt()
+    }
+    url = "https://fhir.epic.com/interconnect-fhir-oauth/oauth2/token"
+    response = requests.post(url, data=data, headers=headers)
+    if response.status_code == 200:
+        return response.json().get('access_token')
+    elif response.status_code >= 400:
+        return response.raise_for_status()
+
+
+def get_patient_fhir_id(given_name, family_name, birthdate, access_token):
+    url = f"https://fhir.epic.com/interconnect-fhir-oauth/api/FHIR/R4/Patient?birthdate={birthdate}&family={family_name}&given={given_name}&_format=json"
+    headers = {
+        'authorization': 'Bearer %s' % access_token,
+    }
+    response = requests.get(url, headers=headers)
+    if response.status_code == 200:
+        response_json = response.json()
+        fhir_id = None
+        entry = response_json.get('entry')[0]
+        if entry: 
+            resource = entry.get('resource')
+            if resource:
+                fhir_id = resource.get('id')
+        return fhir_id
+    elif response.status_code >= 400:
+        response.raise_for_status()
+
+
+# TODO add time param 12 weeks from study start date
+def get_epic_appointments_for_patient(fhir_id, access_token):
+    appointments = []
+    headers = {
+        'authorization': 'Bearer %s' % access_token,
+    }
+    url = f"https://fhir.epic.com/interconnect-fhir-oauth/api/FHIR/R4/Appointment?&patient={fhir_id}&service-category=appointment&_format=json"
+    response = requests.get(url, headers=headers)
+    if response.status_code == 200:
+        json_response = response.json()
+        entries = json_response['entry']
+        for entry in entries:
+            appointments.append(entry)
+    elif response.status_code >= 400:
+        response.raise_for_status()
+    return appointments
+
+
+def convert_date_and_time_to_utc_timestamp(date, time):
+    date_time = date + "T" + time
+    utc_zone = dateutil.tz.gettz('UTC')
+    # Hardcoded for MGH study
+    local_zone = dateutil.tz.gettz('America/New_York')
+    local_datetime = datetime.fromisoformat(date_time)
+    local_datetime = local_datetime.replace(tzinfo=local_zone)
+    utc_datetime = local_datetime.astimezone(utc_zone)
+    utc_iso_format = utc_datetime.strftime('%Y-%m-%dT%H:%M:%SZ')
+
+    return utc_iso_format
+
+
+def convert_utc_timestamp_to_date_and_time(utc_timestamp):
+    utc_zone = dateutil.tz.gettz('UTC')
+    local_zone = dateutil.tz.gettz('America/New_York')
+    utc_datetime = datetime.fromisoformat(utc_timestamp.replace('Z', ''))
+    utc_datetime = utc_datetime.replace(tzinfo=utc_zone)
+    local_datetime = utc_datetime.astimezone(local_zone)
+    date = local_datetime.strftime('%Y-%m-%d')
+    time = local_datetime.strftime('%H:%M')
+
+    return date, time
+
+
+def sync_all_appointments_domain(domain):
```

Comment:
Do you know the kinds of scale we expect (# patient / appointment cases). I'm just wondering if this should be done in batches.

Reply:
I did give some thought to batching. I'm not sure about the number of patients, but I'm estimating that the number of appointments per patient will be relatively low for the timeframe of 12 weeks (still need to add the time constraint to the appointment query). One appointment per week seems like the upper end to me. If you double that it's still only 24 per patient. I'll talk to Chantal and see if she thinks those numbers are low and also get a sense of the number of patients.

Reply:
from Chantal:
_the max number of patients enrolled in the study will be 55, with rolling enrollment over the next year_
 _~5 patients per month and they are enrolled for 12 weeks so maybe ~15 open cases at a time in a given month_